presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java

```diff
@@ -127,6 +168,47 @@ public OrcReader(
         this.cacheable = requireNonNull(cacheable, "hiveFileContext is null");
     }
 
+    private static Map<Integer, Slice> createIntermediateKeysMap(
```

Comment:
Where does presto handle schema mismatch?

day 1: table has struct<a:struct<a:int>, b:float> and we create a partition
day 2: table has new field appended to the first column: struct<a:struct<a:int,b:int>, b:float>, we create another partition

then int -> key map is not stable since same id represents different columns in the partition. Also FBETL has bug so not only that table schema may be different from partition's, partition schema may be different from the partition file. I did not see a place where it's explicitly handled

Reply:
do you mean https://github.com/prestodb/presto/blob/master/presto-orc/src/main/java/com/facebook/presto/orc/AbstractOrcRecordReader.java#L159-L168. 

How is one supposed to map from columns to keys then?

Reply:
I saw this comment before your other one.  I didn't realize the orc column numbers wouldn't be stable (i knew you could add new columns, but didn't realize you could add new fields to row types so that the column numbers would be off). I guess an interface is the way to go

Reply:
So basically i'm not sure about how Presto does schema mismatch. BBIO and DWIO do things differently for example. You solution may be different.

BBIO doesn't handle schema mismatch in the reader, it will always relies on the upper layer to handle it (object inspector craziness in the hive/spark context). So it can always load the node ids without worrying about a new key being added.

DWIO is different in the sense that the reader takes in the table schema, and then tries to match the file schema to it (assuming they are compatible). That step may case file level node id to be changed to match the table, hence id is not stable. So the reader maintains new id to old id map, and does one level of indirection to load the keys.

But regardless i think interface with integer id should be avoided. The contract is too loose and we should not use id to look up the key. In fact, the keyMetadata section stored key reference. BBIO loads all the keys specified in partition metadata to map<keyRef, IEK>, then the reader should just use keyMetadata to look up the actual IEK

Reply:
actually, I'm going to keep this map, but moving it to the record reader constructor.  We will generate the correct column indices before passing it in based on the column handle.  Also, note that the slice here isn't necessarily an intermediate key, it's whatever serialized metadata is required for reading.  The reason we can't just pass in the keyMetadata is because it's not defined that the key is sufficient to determine what the right iek is. if the keyMetadata is just an encrypted key, we'd need the field index to figure out what iek to use to decrypt it.